### DISCUSSION TITLE
How to skip all steps at once?

### DISCUSSION THREAD

**Discussion by evandrocoan** [community/community#26439] [Actions]
<p>When my matrix is not running in a tagged release (only to upload pypi Python 3.8 binary wheels), then, those tests should not run any steps requiring Python 3.8: <a href="https://stackoverflow.com/questions/59759921/how-to-skip-github-actions-job-on-push-event" rel="nofollow noopener">https://stackoverflow.com/questions/59759921/how-to-skip-github-actions-job-on-push-event</a></p>
<p>I tried adding <code>if: matrix.python == '3.7' || startsWith(github.ref, 'refs/tags/')</code> before the <code>steps:</code></p>
<pre><code>jobs:
  tests:
    name: ${{ matrix.name }} ${{ startsWith(github.ref, 'refs/tags/') }} ${{ matrix.BUILD_TYPE }} ${{ matrix.python }}
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        BUILD_TYPE:
          - check
          - build
        os:
          - ubuntu-latest
          - macos-latest
          - windows-latest
        python:
          - 3.7
          - 3.8

    if: matrix.python == '3.7' || startsWith(github.ref, 'refs/tags/')
    steps:
      - run: git config --global core.autocrlf false
</code></pre>
<p>But it is complaining that the <code>matrix</code> variable is not defined. I also tried adding the <code>if</code> inside the <code>steps:</code></p>
<pre><code>steps:
      if: matrix.python == '3.7' || startsWith(github.ref, 'refs/tags/')
      - run: git config --global core.autocrlf false
      - uses: actions/checkout@v2
</code></pre>
<p>But now it complains <code>You have an error in your yaml syntax on the line</code>.</p>
<p>I also tried add an <code>if:</code> to exclude them from the build matrix:</p>
<pre><code>exclude:
          - python: 3.8
            if: !startsWith(github.ref, 'refs/tags/')
            BUILD_TYPE: check

or

        exclude:
          if: !startsWith(github.ref, 'refs/tags/')
          - python: 3.8
            BUILD_TYPE: check

or

        exclude:
          - if: !startsWith(github.ref, 'refs/tags/')
          - python: 3.8
            BUILD_TYPE: check
</code></pre>
<p>But nothing worked either with syntax error or invalid matrix variable.</p>
<p>The only way to fix it was by adding an <code>if: matrix.python == '3.7' || startsWith(github.ref, 'refs/tags/')</code> into every steps of my matrix:</p>
<pre><code>steps:
      - run: git config --global core.autocrlf false
        if: matrix.python == '3.7' || startsWith(github.ref, 'refs/tags/')

      - uses: actions/checkout@v2
        if: matrix.python == '3.7' || startsWith(github.ref, 'refs/tags/')
</code></pre>
<p>But the problem with this is that now every job still being set up and run for Python 3.8 (but the Python 3.8 tests are doing nothing and are being marked as passed). The tests for Python 3.8 should not be run either being marked as passed for doing nothing:</p>
<ol>
<li><a href="https://github.com/evandroforks/anki/runs/634557044?check_suite_focus=true" rel="nofollow noopener">https://github.com/evandroforks/anki/runs/634557044?check_suite_focus=true</a></li>
<li>
<a href="https://github.com/ankitects/anki/pull/599" rel="nofollow noopener">https://github.com/ankitects/anki/pull/599</a> - Set to checks.yml upload Python 3.8 wheels to PyPi</li>
</ol>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/e/ede121205bbb44b14e3402a5f5ca5259c840be85.png" data-download-href="/uploads/short-url/xWnuSm3S5lWVb5pD8hgGW7dntdP.png?dl=1" title="2020-04-30 17_33_50-Set to checks.yml upload Python 3.8 wheels to PyPi · Pull Request.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/5332158/181097838-a3b4c980-574d-41f3-8748-4388bfafe759.png" alt="2020-04-30 17_33_50-Set to checks.yml upload Python 3.8 wheels to PyPi · Pull Request.png" data-base62-sha1="xWnuSm3S5lWVb5pD8hgGW7dntdP" width="563" height="499" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/e/ede121205bbb44b14e3402a5f5ca5259c840be85_2_563x499.png, https://aws1.discourse-cdn.com/github/original/2X/e/ede121205bbb44b14e3402a5f5ca5259c840be85.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/e/ede121205bbb44b14e3402a5f5ca5259c840be85.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/e/ede121205bbb44b14e3402a5f5ca5259c840be85_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">2020-04-30 17_33_50-Set to checks.yml upload Python 3.8 wheels to PyPi · Pull Request.png</span><span class="informations">838×743 67.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by cerna** [comment, marked as answer]
<p>Hi <a class="mention" href="/evandrocoan">@evandrocoan</a>,</p>
<p>can you add one preparatory job at start? Then you could in this job dynamically create a JSON and use the fromJSON function to create the matrix object.</p>
<p>(Described in the <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions" rel="nofollow noopener">tutorial</a>.)</p>

**Comment by evandrocoan** [comment]
<p>Just for completeness, this is an example:</p>
<pre><code>name: build
on: push
jobs:
  job1:
    runs-on: ubuntu-latest
    outputs:
      matrix: ${{ steps.set-matrix.outputs.matrix }}
    steps:
    - id: set-matrix
      run: echo "::set-output name=matrix::{\"include\":[{\"project\":\"foo\",\"config\":\"Debug\"},{\"project\":\"bar\",\"config\":\"Release\"}]}"
  job2:
    needs: job1
    runs-on: ubuntu-latest
    strategy:
      matrix: ${{fromJson(needs.job1.outputs.matrix)}}
    steps:
    - run: build
</code></pre>
<p>And here you can find my full code for this: <a href="https://github.com/ankitects/anki/pull/598/files" rel="nofollow noopener">https://github.com/ankitects/anki/pull/598/files</a></p>
